File: src/drone_teleop/drone_teleop/headless_runner.py
"""Headless MuJoCo runner. Loads scene XML + plugins, steps sim without viewer.

Plugins (AcroMode, OdometryPublisher, ImuPublisher, CollisionPublisher,
ExternalForce) handle ROS2 pub/sub internally — same topics as GUI mode.
"""
import argparse
import os
import signal
import sys
import time

import mujoco


def main():
    p = argparse.ArgumentParser()
    p.add_argument("xml", help="Scene XML path")
    p.add_argument("--plugin-dir", default=None,
                   help="MuJoCo plugin dir (default: <ws>/mujoco-3.4.0/bin/mujoco_plugin)")
    p.add_argument("--realtime", action="store_true",
                   help="Sleep to match wall clock (default: fast-forward)")
    args = p.parse_args()

    ws = os.environ.get("COLCON_UAV_WS_DIR", "/home/bryansgue/uav_ws")
    plugin_dir = args.plugin_dir or os.path.join(ws, "mujoco-3.4.0", "bin", "mujoco_plugin")

    if os.path.isdir(plugin_dir):
        mujoco.mj_loadAllPluginLibraries(plugin_dir)
        print(f"[headless] plugins loaded from {plugin_dir}", flush=True)
    else:
        print(f"[headless] WARN plugin dir not found: {plugin_dir}", flush=True)

    m = mujoco.MjModel.from_xml_path(args.xml)
    d = mujoco.MjData(m)
    dt = m.opt.timestep
    print(f"[headless] model loaded, dt={dt}s, nq={m.nq}, nbody={m.nbody}", flush=True)

    stop = {"flag": False}

    def _sig(_a, _b):
        stop["flag"] = True
    signal.signal(signal.SIGINT, _sig)
    signal.signal(signal.SIGTERM, _sig)

    t0 = time.time()
    sim_t = 0.0
    next_log = 1.0
    while not stop["flag"]:
        mujoco.mj_step(m, d)
        sim_t += dt
        if args.realtime:
            wall = time.time() - t0
            if sim_t > wall:
                time.sleep(sim_t - wall)
        if sim_t >= next_log:
            next_log += 5.0
            wall = time.time() - t0
            rtf = sim_t / wall if wall > 0 else 0.0
            print(f"[headless] sim_t={sim_t:.1f}s wall={wall:.1f}s RTF={rtf:.2f}", flush=True)

    print("[headless] shutdown", flush=True)
    sys.exit(0)


if __name__ == "__main__":
    main()

File: src/MujocoRosUtils/plugin/headless_runner.cpp
// Headless MuJoCo runner. Loads scene XML + plugins, steps sim without viewer.
// Plugins (AcroMode, OdometryPublisher, ImuPublisher, CollisionPublisher,
// ExternalForce) handle ROS2 pub/sub internally.
#include <mujoco/mujoco.h>

#include <atomic>
#include <chrono>
#include <csignal>
#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <string>
#include <thread>

namespace
{
std::atomic<bool> g_stop{false};
void on_sig(int) { g_stop.store(true); }
}  // namespace

int main(int argc, char ** argv)
{
  const char * xml_path = nullptr;
  const char * plugin_dir = nullptr;
  bool realtime = false;

  for(int i = 1; i < argc; ++i)
  {
    if(std::strcmp(argv[i], "--plugin-dir") == 0 && i + 1 < argc) { plugin_dir = argv[++i]; }
    else if(std::strcmp(argv[i], "--realtime") == 0) { realtime = true; }
    else if(argv[i][0] != '-') { xml_path = argv[i]; }
  }

  if(!xml_path)
  {
    std::fprintf(stderr, "usage: headless_runner <scene.xml> [--plugin-dir DIR] [--realtime]\n");
    return 1;
  }

  if(plugin_dir)
  {
    mj_loadAllPluginLibraries(plugin_dir, nullptr);
    std::printf("[headless] plugins loaded from %s\n", plugin_dir);
    std::fflush(stdout);
  }

  char err[1024] = {0};
  mjModel * m = mj_loadXML(xml_path, nullptr, err, sizeof(err));
  if(!m)
  {
    std::fprintf(stderr, "[headless] mj_loadXML failed: %s\n", err);
    return 2;
  }
  mjData * d = mj_makeData(m);
  const double dt = m->opt.timestep;
  std::printf("[headless] model loaded, dt=%.4fs, nq=%d, nbody=%d\n", dt, m->nq, m->nbody);
  std::fflush(stdout);

  std::signal(SIGINT, on_sig);
  std::signal(SIGTERM, on_sig);

  using clk = std::chrono::steady_clock;
  const auto t0 = clk::now();
  double sim_t = 0.0;
  double next_log = 5.0;

  while(!g_stop.load())
  {
    mj_step(m, d);
    sim_t += dt;

    if(realtime)
    {
      const double wall = std::chrono::duration<double>(clk::now() - t0).count();
      if(sim_t > wall)
      {
        std::this_thread::sleep_for(std::chrono::duration<double>(sim_t - wall));
      }
    }

    if(sim_t >= next_log)
    {
      next_log += 5.0;
      const double wall = std::chrono::duration<double>(clk::now() - t0).count();
      const double rtf = wall > 0 ? sim_t / wall : 0.0;
      std::printf("[headless] sim_t=%.1fs wall=%.1fs RTF=%.2f\n", sim_t, wall, rtf);
      std::fflush(stdout);
    }
  }

  mj_deleteData(d);
  mj_deleteModel(m);
  std::printf("[headless] shutdown\n");
  return 0;
}
